src/docs/components/reactive-data.md

```diff
@@ -4,78 +4,208 @@ description: Reactive Data, Handling arrays and objects
 
 ```tsx
-import { Prop, State, Watch } from '@stencil/core';
+// We import Prop & State to show how `@Watch()` can be used on
+// class members decorated with either `@Prop()` or `@State()`
+import { Component, Prop, State, Watch } from '@stencil/core';
 
+@Component({
+  tag: 'loading-indicator' 
+})
 export class LoadingIndicator {
+  // We decorate a class member with @Prop() so that we
+  // can apply @Watch()
   @Prop() activated: boolean;
+  // We decorate a class member with @State() so that we
+  // can apply @Watch()
   @State() busy: boolean;
 
+  // Apply @Watch() for the component's `activated` member.
+  // Whenever `activated` changes, this method will fire.
   @Watch('activated')
   watchPropHandler(newValue: boolean, oldValue: boolean) {
+    console.log('The old value of activated is: ', oldValue);
     console.log('The new value of activated is: ', newValue);
   }
-  
+
+  // Apply @Watch() for the component's `busy` member.
+  // Whenever `busy` changes, this method will fire.
   @Watch('busy')
   watchStateHandler(newValue: boolean, oldValue: boolean) {
+    console.log('The old value of busy is: ', oldValue);
     console.log('The new value of busy is: ', newValue);
   }
 }
 ```
+In the example above, there are two `@Watch()` decorators.
+One decorates `watchPropHandler`, which will fire when the class member `activated` changes.
+The other decorates `watchStateHandler`, which will fire when the class member `busy` changes.
+
+When fired, the `@Watch()`'ed method will receive the old and new values of the prop/state.
+This is useful for validation or the handling of side effects.
+
+> The `@Watch()` decorator does not fire when a component initially loads.
 
+## Handling Arrays and Objects
 
-## Handling arrays and objects
+When Stencil checks if a class member decorated with `@Prop()` or `@State()` has changed, it checks if the reference to the class member has changed.
+When a class member is an object or array, and is marked with `@Prop()` or `@State`, in-place mutation of an existing entity will _not_ cause `@Watch()` to fire, as it does not change the _reference_ to the class member.
 
-To update array or object data, use the following techniques, which are fast-becoming a core part of the modern JavaScript  toolbox.
+### Updating Arrays
 
-### Updating arrays
+For arrays, the standard mutable array operations such as `push()` and `unshift()` won't trigger a component update.
+These functions will change the content of the array, but won't change the reference to the array itself.
 
-For arrays, the standard mutable array operations such as `push()` and `unshift()` won't trigger a component update. Instead, non-mutable array operators should be used as they return a copy of a new array. These include `map()` and `filter()`, and the ES6 spread operator syntax.
+In order to make changes to an array, non-mutable array operators should be used.
```

Comment:
would it be potentially clearer to say something like "the prop or state member should be assigned to a new array, created using a method like `.map()` or `filter()`, the spread operator syntax, or simply a new array literal" ? I wouldn't want to potentially give the impression that just calling `.map` will be enough (the difference being that you need to assign the new array which is the return value of map).

Reply:
yup, I think it's good :)